sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/CosmosDBSourceConnector.java

```diff
@@ -0,0 +1,347 @@
+    }
+
+    @Override
+    public List<Map<String, String>> taskConfigs(int maxTasks) {
+        // For now, we start with copying data by feed range
+        // but in the future, we can have more optimization based on the data size etc.
+        return this.getTaskConfigs(maxTasks);
+    }
+
+    @Override
+    public void stop() {
+        logger.info("Stopping Kafka CosmosDB source connector");
+        if (this.cosmosClient != null) {
+            logger.debug("Closing cosmos client");
+            this.cosmosClient.close();
+        }
+
+        if (this.monitorThread != null) {
+            logger.debug("Closing monitoring thread");
+            this.monitorThread.close();
+        }
+    }
+
+    @Override
+    public ConfigDef config() {
+        return CosmosSourceConfig.getConfigDef();
+    }
+
+    @Override
+    public String version() {
+        return CosmosConstants.currentVersion;
+    } // TODO: how this is being used
+
+    private List<Map<String, String>> getTaskConfigs(int maxTasks) {
+        Pair<MetadataTaskUnit, List<FeedRangeTaskUnit>> taskUnits = this.getAllTaskUnits();
+
+        // The metadataTaskUnit is a one time only task when the connector starts/restarts,
+        // so there is no need to assign a dedicated task thread for it
+        // we are just going to assign it to one of the tasks which processing feedRanges tasks
+        List<List<FeedRangeTaskUnit>> partitionedTaskUnits = new ArrayList<>();
+        if (taskUnits.getRight().size() <= maxTasks) {
+            partitionedTaskUnits.addAll(
+                taskUnits.getRight().stream().map(taskUnit -> Arrays.asList(taskUnit)).collect(Collectors.toList()));
+        } else {
+            // using round-robin fashion to assign tasks to each buckets
+            for (int i = 0; i < maxTasks; i++) {
+                partitionedTaskUnits.add(new ArrayList<>());
+            }
+
+            for (int i = 0; i < taskUnits.getRight().size(); i++) {
+                partitionedTaskUnits.get(i % maxTasks).add(taskUnits.getRight().get(i));
+            }
+        }
+
+        List<Map<String, String>> allSourceTaskConfigs = new ArrayList<>();
+        partitionedTaskUnits.forEach(feedRangeTaskUnits -> {
+            Map<String, String> taskConfigs = this.config.originalsStrings();
+            taskConfigs.putAll(
+                CosmosSourceTaskConfig.getFeedRangeTaskUnitsConfigMap(feedRangeTaskUnits));
+            allSourceTaskConfigs.add(taskConfigs);
+        });
+
+        // assign the metadata task to the last of the task config as it has least number of feedRange task units
+        allSourceTaskConfigs
+            .get(allSourceTaskConfigs.size() - 1)
+            .putAll(CosmosSourceTaskConfig.getMetadataTaskUnitConfigMap(taskUnits.getLeft()));
+
+        return allSourceTaskConfigs;
+    }
+
+    private Pair<MetadataTaskUnit, List<FeedRangeTaskUnit>> getAllTaskUnits() {
+        // TODO: add transient errors handling
```

Comment:
Can we add a suffix for TODOs indicating whether the TODO needs to be addressed - before merging PR, before public Beta, before GA?

Reply:
based on the offline discussion, will reply on the retry in azure-cosmos, no customized retries will be added here. Only change is to convert into RetriableException if it is transient exceptions.